### PR TITLE
[HttpFoundation] Add type-hints

### PR DESCRIPTION
**Pull request by julien57**
| Q             | A
| ------------- | ---
| Branch?       | master 
| Bug fix?      | no
| New feature?  | no 
| BC breaks?    | no     
| Deprecations? | no 
| Tests pass?   | yes    
| Fixed tickets | #32179  
| License       | MIT
| Doc PR        |  

First commit to add type-hints to HttpFoundation

### CODE REVIEW

File: src/Symfony/Component/HttpFoundation/AcceptHeader.php [hunk 1]
```diff
@@ -48,11 +48,11 @@ public function __construct(array $items)
      *
      * @return self
      */
-    public static function fromString($headerValue)
+    public static function fromString(string $headerValue)
```
Comment (derrabus) [inline src/Symfony/Component/HttpFoundation/AcceptHeader.php]: `@param` can be removed from multiple docblocks.

File: src/Symfony/Component/HttpFoundation/Cookie.php [hunk 2]
```diff
@@ -68,6 +68,19 @@ public static function fromString($cookie, $decode = false)
...
+     * @param int         $expire
+     * @param string|null $path
+     * @param string|null $domain
+     * @param bool|null   $secure
+     * @param bool        $httpOnly
+     * @param bool        $raw
+     * @param string|null $sameSite
+     *
+     * @return self
+     */
```
Comment (derrabus) [inline src/Symfony/Component/HttpFoundation/Cookie.php]: This docblock does not add any new information. Please remove it.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: .gitignore [hunk 3]
```diff
@@ -1,8 +0,0 @@
-vendor/
-composer.lock
-phpunit.xml
-.php_cs.cache
-composer.phar
-package.tar
-/packages.json
-/.phpunit
```
Comment (derrabus) [inline .gitignore]: Please don‘t delete that file.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/HttpFoundation/File/MimeType/ExtensionGuesserInterface.php [hunk 4]
```diff
@@ -27,5 +27,5 @@ interface ExtensionGuesserInterface
      *
      * @return string The guessed extension or NULL, if none could be guessed
      */
-    public function guess($mimeType);
+    public function guess(string $mimeType);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpFoundation/File/MimeType/ExtensionGuesserInterface.php]: all classes in the MimeType subnamespace are deprecated: they should be removed instead

**Comment by nicolas-grekas** [comment]
@julien57ok to finish this one?

File: src/Symfony/Component/HttpFoundation/AcceptHeader.php [hunk 5]
```diff
@@ -78,23 +76,19 @@ public function __toString()
...
 
     /**
      * Returns given value's item, if exists.
      *
-     * @param string $value
-     *
      * @return AcceptHeaderItem|null
      */
-    public function get($value)
+    public function get(string $value)
```
Comment (xuanquynh) [inline src/Symfony/Component/HttpFoundation/AcceptHeader.php]: Should we use this code?

```php
public function get(string $value) ?AcceptHeaderItem
```

Comment (derrabus) [reply src/Symfony/Component/HttpFoundation/AcceptHeader.php]: 👎  Introducing a return type to a non-final method is a breaking change.

Comment (xuanquynh) [reply src/Symfony/Component/HttpFoundation/AcceptHeader.php]: This PR is targeted to master branch, I think breaking changes are valid if they are really useful. It's just my question for more details.

Comment (xuanquynh) [reply src/Symfony/Component/HttpFoundation/AcceptHeader.php]: In fact, I miss the case "Don't add return types. They require a different strategy.".

**Comment by nicolas-grekas** [comment]
I just push-forced on this PR, review welcome.

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 6]
```diff
@@ -384,19 +383,13 @@ public function send()
...
      */
-    public function setContent($content)
+    public function setContent(?string $content)
     {
-        if (null !== $content && !\is_string($content) && !is_numeric($content) && !\is_callable([$content, '__toString'])) {
-            throw new \UnexpectedValueException(sprintf('The Response content must be a string or object implementing __toString(), "%s" given.', \gettype($content)));
-        }
-
-        $this->content = (string) $content;
+        $this->content = $content;
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Response.php]: since `$content` is nullable, `$this->content` can also be null. this might break for people expecting string in `getContent`. so I guess we need to cast null to `''`

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 7]
```diff
@@ -209,11 +209,10 @@ public function __construct($content = '', int $status = 200, array $headers = [
      *
      * @param mixed $content The response content, see setContent()
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Response.php]: outdated mixed param

File: src/Symfony/Component/HttpFoundation/Response.php [hunk 8]
```diff
@@ -384,19 +383,13 @@ public function send()
...
-     *
      * @return $this
      *
      * @throws \UnexpectedValueException
      */
-    public function setContent($content)
+    public function setContent(?string $content)
     {
-        if (null !== $content && !\is_string($content) && !is_numeric($content) && !\is_callable([$content, '__toString'])) {
-            throw new \UnexpectedValueException(sprintf('The Response content must be a string or object implementing __toString(), "%s" given.', \gettype($content)));
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Response.php]: fully agree as the casting should be done in user-land when they need strict types compatibility.

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 9]
```diff
@@ -1586,8 +1553,6 @@ public function getPreferredFormat(?string $default = 'html'): ?string
     /**
      * Returns the preferred language.
      *
-     * @param array $locales An array of ordered available locales
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Request.php]: I think we should document this as `@param string[] $locales`. otherwise the return type would also not work.

File: src/Symfony/Component/HttpFoundation/Request.php [hunk 10]
```diff
@@ -1296,11 +1278,9 @@ public static function getMimeTypes($format)
     /**
      * Gets the format associated with the mime type.
      *
-     * @param string $mimeType The associated mime type
-     *
      * @return string|null The format (null if not found)
      */
-    public function getFormat($mimeType)
+    public function getFormat(?string $mimeType)
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/Request.php]: making this nullable seems strange. I don't see the use-case for this.

File: src/Symfony/Component/HttpFoundation/RedirectResponse.php [hunk 11]
```diff
@@ -74,13 +70,11 @@ public function getTargetUrl()
...
-     * @param string $url The URL to redirect to
-     *
      * @return $this
      *
      * @throws \InvalidArgumentException
      */
-    public function setTargetUrl($url)
+    public function setTargetUrl(string $url)
     {
         if ('' === ($url ?? '')) {
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/RedirectResponse.php]: the code suggests `$url` can be null, but it's not nullable at the moment.

File: src/Symfony/Component/HttpFoundation/HeaderBag.php [hunk 12]
```diff
@@ -197,14 +179,11 @@ public function remove($key)
...
      *
-     * @param string    $key     The parameter key
-     * @param \DateTime $default The default value
-     *
      * @return \DateTime|null The parsed DateTime or the default value if the header does not exist
      *
      * @throws \RuntimeException When the HTTP header is not parseable
      */
-    public function getDate($key, \DateTime $default = null)
+    public function getDate(string $key, \DateTime $default = null)
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/HeaderBag.php]: we should probably change \DateTime to \DateTimeInterface and adjust the return phpdoc as well. but that could be a separate PR.

File: src/Symfony/Component/HttpFoundation/AcceptHeaderItem.php [hunk 13]
```diff
@@ -173,12 +162,9 @@ public function getAttributes()
     /**
      * Set an attribute.
      *
-     * @param string $name
-     * @param string $value
-     *
      * @return $this
      */
-    public function setAttribute($name, $value)
+    public function setAttribute(string $name, string $value)
```
Comment (Tobion) [inline src/Symfony/Component/HttpFoundation/AcceptHeaderItem.php]: there is still a `(string) $value` cast left in the implementation

**Review by Tobion — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/HttpFoundation/Response.php (hunk 6 above)
Comment (Tobion) [reply src/Symfony/Component/HttpFoundation/Response.php]: there is actually a failing test due to this: https://travis-ci.org/symfony/symfony/jobs/569486612

**Comment by nicolas-grekas** [comment]
Continued in #33088, thanks @julien57
